lib/SIL/SILFunction.cpp

```diff
@@ -162,11 +162,15 @@ bool SILFunction::shouldOptimize() const {
 }
 
 Type SILFunction::mapTypeIntoContext(Type type) const {
+  assert((!type->hasTypeParameter() || getGenericEnvironment()) &&
+         "SILFunction should have a generic environment");
   return GenericEnvironment::mapTypeIntoContext(
       getGenericEnvironment(), type);
 }
 
 SILType SILFunction::mapTypeIntoContext(SILType type) const {
+  assert((!type.hasTypeParameter() || getGenericEnvironment()) &&
```

Comment:
They do not assert that the type is fully concrete if the environment is null, AFAIK:

```c++
Type GenericEnvironment::mapTypeIntoContext(GenericEnvironment *env,
                                            Type type) {
  assert(!type->hasArchetype() && "already have a contextual type");

  if (!env)
    return type.substDependentTypesWithErrorTypes();

  return env->mapTypeIntoContext(type);
}

Type
GenericEnvironment::mapTypeOutOfContext(GenericEnvironment *env,
                                        Type type) {
  assert(!type->hasTypeParameter() && "already have an interface type");

  if (!env)
    return type.substDependentTypesWithErrorTypes();

  return env->mapTypeOutOfContext(type);
}
```